### PR TITLE
Trestle import and unit testing

### LINKED ISSUE #38: Implement `trestle import`
## Issue description / feature objectives

Implement `trestle import` according to [specifications](https://github.com/IBM/compliance-trestle/blob/develop/docs/trestle-spec.md).

## Completion Criteria

Complete implementation and testing.

## Expected Behavior

- Contents of OSCAL model in imported file decomposed and placed in the expected folder as specification.

### PR DESCRIPTION
**Pull request by jeffdmgit**
## Types of changes

- \[ \] Bug fix (non-breaking change which fixes an issue)
- \[x\] New feature (non-breaking change which adds functionality)
- \[ \] Breaking change (fix or feature that would cause existing functionality to change)
- \[x\] My code follows the code style of this project.
- \[ \] My change requires a change to the documentation.
- \[ \] I have updated the documentation accordingly.
- \[x\] I have added tests to cover my changes.
- \[x\] All new and existing tests passed.
- \[ \] All commits are signed.

This closes #38.

### CODE REVIEW

**Comment by jeffdmgit** [comment]
Failing import__test.py in Windows, it seems, when importing a target definition. Otherwise checks fine on Linux and MacOS, although macos-latest, 3.8 was cancelled, probably when the Windows checks failed.

**Review by butler54 — CHANGES_REQUESTED** [review]
The test suite is a bit messy and can be cleaned up. The code in the actual function looks pretty good (+- the error handling which is in the messages).

Once the test suite is a bit cleaner ping me and I can re-review.

File: trestle/core/commands/import_.py [hunk 1]
```diff
@@ -14,11 +14,147 @@
...
+            logger.error(f'Current working directory: {cwd} is not within a trestle project.')
+            return 1
+
+        # 2. Importing a file that is already inside a trestle-initialized dir is bad
+        trestle_root = trestle_root.resolve()
+        try:
+            input_file.absolute().relative_to(trestle_root)
+        except Exception:
+            # An exception here is good: it means that the input file is not inside a trestle dir.
+            pass
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply trestle/core/commands/import_.py]: Fixed for all occurrences.

File: trestle/core/commands/import_.py [hunk 2]
```diff
@@ -14,11 +14,147 @@
...
+        else:
+            logger.error('Input file cannot be from current trestle project. Use duplicate instead.')
+            return 1
+
+        # 3. Work out typing information from input suffix.
+        try:
+            content_type = FileContentType.to_content_type(input_file.suffix)
+        except Exception as err:
+            logger.debug(f'FileContentType.to_content_type() failed: {err}')
+            logger.error(f'Import failed (FileContentType.to_content_type()): {err}')
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply trestle/core/commands/import_.py]: Provided clearer error logs. Retained references to internal classes in debug logs.

File: trestle/core/commands/import_.py [hunk 3]
```diff
@@ -14,11 +14,147 @@
...
+            return 1
+
+        # 4.3 parse the model
+        parent_model_name = parser.to_full_model_name(parent_alias)
+        try:
+            parent_model = parser.parse_file(input_file.absolute(), parent_model_name)
+        except Exception as err:
+            logger.debug(f'parser.parse_file() failed: {err}')
+            logger.error(f'Import failed (parser.parse_file()): {err}')
+            return 1
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply trestle/core/commands/import_.py]: Ok. More thinking via issue #256 then.

File: trestle/core/commands/import_.py [hunk 4]
```diff
@@ -14,11 +14,147 @@
...
+
+        # 5. Work out output directory and file
+        plural_path: str
+        plural_path = parent_alias
+        # Cater to POAM
+        if parent_alias[-1] != 's':
+            plural_path = parent_alias + 's'
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply trestle/core/commands/import_.py:124]: Handle via #256?

File: tests/trestle/core/commands/import__test.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    test_args = f'trestle import -f {str(catalog_file_path)} -o catalog'.split()
+    with patch('trestle.utils.fs.get_trestle_project_root') as get_trestle_project_root_mock:
+        get_trestle_project_root_mock.return_value = None
+        with patch.object(sys, 'argv', test_args):
+            try:
+                Trestle().run()
+            except Exception:
+                assert True
+            else:
+                AssertionError()
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py]: Simplified with return code checking. No exceptions are raised on the import code being tested anyway.

File: tests/trestle/core/commands/import__test.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    json_file = tempfile.NamedTemporaryFile(suffix='.json')
+    sample_data = '"star": {'
+    json_file.write(sample_data.encode('utf8'))
+    json_file.seek(0)
+    test_args = f'trestle import -f {str(json_file.name)} -o imported'.split()
+    with patch('trestle.utils.fs.load_file') as load_file_mock:
+        load_file_mock.side_effect = err.TrestleError('stuff')
+        with patch.object(sys, 'argv', test_args):
+            rc = Trestle().run()
+            assert rc == 1
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py:156]: Not quite following. I think we ought to let the full Trestle() class run lest we miss the initial args validations and what not.

File: tests/trestle/core/commands/import__test.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    test_args = f'trestle import -f {sample_file.name} -o catalog'.split()
+    with patch('trestle.core.parser.parse_file') as parse_file_mock:
+        parse_file_mock.side_effect = err.TrestleError('stuff')
+        with patch.object(sys, 'argv', test_args):
+            try:
+                Trestle().run()
+            except Exception:
+                assert True
+            else:
+                AssertionError()
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py]: Simplified to checking for the return code.

File: tests/trestle/core/commands/import__test.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    # This seek is necessary to flush to file.
+    sample_file.seek(0)
+    test_args = f'trestle import -f {sample_file.name} -o catalog'.split()
+    with patch.object(sys, 'argv', test_args):
+        try:
+            Trestle().run()
+        except Exception:
+            assert True
+        else:
+            AssertionError()
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py]: Done.

File: trestle/core/commands/import_.py [hunk 9]
```diff
@@ -14,11 +14,147 @@
...
+            return 1
+
+        try:
+            import_plan.execute()
+        except Exception as err:
+            logger.debug(f'import_plan.execute() failed: {err}')
+            logger.error(f'Import failed (import_plan.execute()): {err}')
+            return 1
+
+        # 7. Leave the rest to trestle split
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply trestle/core/commands/import_.py:160]: Not yet, at least not beyond splitting down one level, from my testing. Best to leave it until split can definitely process deeper down element level paths

File: tests/trestle/core/commands/import__test.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        }
+    }
+    catalog_file.write(json.dumps(sample_data).encode('utf8'))
+    catalog_file.seek(0)
+    test_args = f'trestle import -f {str(catalog_file.name)} -o imported'.split()
+    with patch('trestle.core.models.plans.Plan.simulate') as simulate_plan_mock:
+        simulate_plan_mock.side_effect = err.TrestleError('stuff')
+        with patch.object(sys, 'argv', test_args):
+            rc = Trestle().run()
+            assert rc == 1
```
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py:234]: Added the test with an attempt to import `groups`, but the response from *parser.parse_file()* isn't entirely specific. It just says 'model_name is required'. This is the same response if one attempts to import valid json with a key that isn't a valid element at all, e.g., _{ "id": "000" }_. If there a need to distinguish, then it has to be done from _parser.parse_file()_.

[the remark this answers is not in the document]
Comment (butler54) [reply tests/trestle/core/commands/import__test.py:255]: Yes I agree.

File: tests/trestle/core/commands/import__test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (butler54) [reply tests/trestle/core/commands/import__test.py:168]: So I think there are two usescases.

The first is testing 'did the class actually receive all the mandatory arguments. E.g. we are testing whether we are getting the expected behaviour from ilcli and argparse. This should be done with Trestle.run().

For testing for a specific behaviour for the unit test we should be testing as tight as possible in terms of the unit test. This means we have clarity that nothing is capturing exceptions where we dont expect (say for example in ILCLI).

See as an example this create command test below (it does not call `_run` for some esoteric reasons, however, the  example holds true.

```python
def test_fail_overwrite(tmp_trestle_dir: pathlib.Path) -> None:
    """Test that a failure occurs when doubling up a create."""
    args = argparse.Namespace(extension='json', name='my_catalog', verbose=0)
    rc = create.CreateCmd.create_object('catalog', Catalog, args)
    assert rc == 0
    rc = create.CreateCmd.create_object('catalog', Catalog, args)
    assert rc > 0

```

File: tests/trestle/core/commands/import__test.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+        'catalog': {
+            'uuid': 'ad0d0a7c-9634-48d9-ba90-fd10bcaf45b8',
+            'metadata': {
+                'title': 'Generic catalog created by trestle.',
+                'last-modified': '2020-12-07T06:18:18.430+00:00',
+                'version': '0.0.0',
+                'oscal-version': 'v1.0.0-milestone3'
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (butler54) [reply tests/trestle/core/commands/import__test.py:41]: @jeffdmgit you missed this:?

**Review by butler54 — CHANGES_REQUESTED** [review]
See the unclosed comments. I think there is a little way to go but it's close.

File: tests/trestle/core/commands/import__test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py]: Done! I retained the happy path test calling Trestle().run() while the rest call _run() on an instance.

**Comment by jeffdmgit** [comment]
> See the unclosed comments. I think there is a little way to go but it's close.

Please review again.

**Review by fsuits — APPROVED** [review]
LGTM

**Review by jeffdmgit — COMMENTED** [review]
I used generators except in cases where I needed to attempt an import of invalid data.

File: tests/trestle/core/commands/import__test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeffdmgit) [reply tests/trestle/core/commands/import__test.py]: Already done.

**Review by butler54 — APPROVED** [review]
LGTM

**Pull request merged**